packages/app-check/src/internal-api.test.ts

```diff
@@ -294,75 +361,75 @@ describe('internal api', () => {
+    it('notifies the listener with the valid token in memory immediately', async () => {
       const clock = useFakeTimers();
-      const fakeListener = (token: AppCheckTokenResult): void => {
-        expect(token).to.deep.equal({
-          token: `fake-memory-app-check-token`
-        });
-        clock.restore();
-        done();
-      };
+      const listener = stub();
 
       setState(app, {
         ...getState(app),
         token: {
           token: `fake-memory-app-check-token`,
-          expireTimeMillis: 123,
+          expireTimeMillis: Date.now() + 60000,
           issuedAtTimeMillis: 0
         }
       });
 
-      addTokenListener(app, fakePlatformLoggingProvider, fakeListener);
+      addTokenListener(app, fakePlatformLoggingProvider, listener);
+      await clock.runAllAsync();
+      expect(listener).to.be.calledWith({
+        token: 'fake-memory-app-check-token'
+      });
+      clock.restore();
     });
 
     it('notifies the listener with the valid token in storage', done => {
-      const clock = useFakeTimers();
-      activate(app, FAKE_SITE_KEY, true);
-      stub(storage, 'readTokenFromStorage').returns(
-        Promise.resolve({
-          token: `fake-cached-app-check-token`,
-          expireTimeMillis: 123,
-          issuedAtTimeMillis: 0
-        })
-      );
+      activate(app, FAKE_SITE_KEY);
+      storageReadStub.resolves({
+        token: `fake-cached-app-check-token`,
+        expireTimeMillis: Date.now() + 60000,
+        issuedAtTimeMillis: 0
+      });
 
+      // Need to use done() if the callback will be called by the
+      // refresher.
       const fakeListener = (token: AppCheckTokenResult): void => {
         expect(token).to.deep.equal({
           token: `fake-cached-app-check-token`
         });
-        clock.restore();
         done();
       };
 
       addTokenListener(app, fakePlatformLoggingProvider, fakeListener);
-      clock.tick(1);
     });
 
-    it('notifies the listener with the debug token immediately', done => {
-      const fakeListener = (token: AppCheckTokenResult): void => {
-        expect(token).to.deep.equal({
-          token: `my-debug-token`
-        });
-        done();
-      };
+    it('notifies the listener with the debug token immediately', async () => {
+      stub(client, 'exchangeToken').resolves(fakeRecaptchaAppCheckToken);
+      const listener = stub();
+
+      const clock = useFakeTimers();
 
       const debugState = getDebugState();
       debugState.enabled = true;
       debugState.token = new Deferred();
       debugState.token.resolve('my-debug-token');
 
-      activate(app, FAKE_SITE_KEY, true);
-      addTokenListener(app, fakePlatformLoggingProvider, fakeListener);
+      activate(app, FAKE_SITE_KEY, false);
+      addTokenListener(app, fakePlatformLoggingProvider, listener);
+      await clock.runAllAsync();
+      expect(listener).to.be.calledWith({ token: 'my-debug-token' });
```

Comment:
The listener should get `fakeRecaptchaAppCheckToken` instead of `my-debug-token`. The [logic](https://github.com/firebase/firebase-js-sdk/blob/master/packages/app-check/src/internal-api.ts#L189-L198) in `addTokenListener` is wrong. (I think initially we thought the debug token can be used directly without exchanging for an appcheck token)

There need not to be special handling for the debug mode, and we should always create a token refresher.

Reply:
Oh oops, thanks for the catch. Fixed `addTokenListener()` and removed these cases.